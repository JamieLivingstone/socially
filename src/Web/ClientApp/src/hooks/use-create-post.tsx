import axios from 'axios';
import { useMutation } from 'react-query';
import { useNavigate } from 'react-router-dom';

import { useAuth } from './use-auth';

export type CreatePostResponse = {
  slug: string;
};

export function useCreatePost() {
  const navigate = useNavigate();
  const { account } = useAuth();

  const { data, isLoading, isError, error, mutateAsync } = useMutation(
    async function createPostRequest(payload: { title: string; body: string; tags: Array<string> }) {
      const { data } = await axios.post<CreatePostResponse>('/api/v1/posts', payload);

      navigate(`/${account?.username}/${data.slug}`);
    },
    { retry: false },
  );

  return {
    data,
    isLoading,
    isError,
    error,
    mutateAsync,
  };
}
