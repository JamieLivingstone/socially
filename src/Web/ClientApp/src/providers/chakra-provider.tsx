import { ChakraProvider as Provider, Theme, extendTheme, withDefaultColorScheme } from '@chakra-ui/react';
import React, { ReactNode } from 'react';

const theme = extendTheme(withDefaultColorScheme({ colorScheme: 'green' }), {
  initialColorMode: 'light',
  useSystemColorMode: false,
  styles: {
    global: ({ theme }: { theme: Theme }) => ({
      body: {
        bg: theme.colors.gray['100'],
      },
    }),
  },
});

export function ChakraProvider({ children }: { children: ReactNode }) {
  return <Provider theme={theme}>{children}</Provider>;
}
