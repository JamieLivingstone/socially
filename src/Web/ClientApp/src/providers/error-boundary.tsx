import axios from 'axios';
import React, { Component } from 'react';

import { Error } from '../screens/miscellaneous/error';
import { NotFound } from '../screens/miscellaneous/not-found';

type ErrorBoundaryProps = {
  children: React.ReactNode;
};

type ErrorBoundaryState = {
  hasError: boolean;
  error: Error | null;
};

export class ErrorBoundary extends Component<ErrorBoundaryProps, ErrorBoundaryState> {
  constructor(props: ErrorBoundaryProps) {
    super(props);
    this.state = { hasError: false, error: null };
  }

  static getDerivedStateFromError(error: Error) {
    return { hasError: true, error };
  }

  render() {
    if (this.state.hasError) {
      if (axios.isAxiosError(this.state.error) && this.state.error.response?.status === 404) {
        return <NotFound />;
      }

      return <Error />;
    }

    return this.props.children;
  }
}
