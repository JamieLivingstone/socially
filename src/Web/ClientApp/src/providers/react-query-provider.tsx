import axios from 'axios';
import React, { ReactNode } from 'react';
import { QueryClient, QueryClientProvider } from 'react-query';

const RETRY_STATUS_CODES = [408, 413, 429, 500, 502, 503, 504];

const client = new QueryClient({
  defaultOptions: {
    queries: {
      refetchOnWindowFocus: false,
      retry: (failureCount, error) => {
        if (
          axios.isAxiosError(error) &&
          error.response &&
          failureCount < 3 &&
          RETRY_STATUS_CODES.includes(error.response.status)
        ) {
          return true;
        }

        return false;
      },
    },
  },
});

export function ReactQueryProvider({ children }: { children: ReactNode }) {
  return <QueryClientProvider client={client}>{children}</QueryClientProvider>;
}
