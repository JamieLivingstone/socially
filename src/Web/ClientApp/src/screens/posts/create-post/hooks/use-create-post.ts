import { useToast } from '@chakra-ui/react';
import axios from 'axios';
import { useMutation } from 'react-query';
import { useNavigate } from 'react-router-dom';

import { useAuth } from '@hooks/use-auth';

export type CreatePostResponse = {
  slug: string;
};

export function useCreatePost() {
  const { account } = useAuth();
  const navigate = useNavigate();
  const toast = useToast();

  const { mutateAsync } = useMutation(
    async function createPostRequest(payload: { title: string; body: string; tags: Array<string> }) {
      try {
        const { data } = await axios.post<CreatePostResponse>('/api/v1/posts', payload);

        navigate(`/${account?.username}/${data.slug}`);
      } catch (error) {
        toast({
          title: 'Failed to create post!',
          status: 'error',
          duration: 3000,
          isClosable: false,
        });
      }
    },
    { useErrorBoundary: false },
  );

  return {
    createPost: mutateAsync,
  };
}
