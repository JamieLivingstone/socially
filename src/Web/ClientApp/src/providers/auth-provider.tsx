import React, { ReactNode } from 'react';

import { Loading } from '../components';
import { AuthContext, useProvideAuth } from '../hooks';

export function AuthProvider({ children }: { children: ReactNode }) {
  const auth = useProvideAuth();

  if (auth.initializing) {
    return <Loading />;
  }

  return <AuthContext.Provider value={auth}>{children}</AuthContext.Provider>;
}
