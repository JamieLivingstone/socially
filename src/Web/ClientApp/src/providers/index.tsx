import React, { ReactNode } from 'react';
import { BrowserRouter } from 'react-router-dom';

import { AuthProvider } from './auth-provider';
import { ChakraProvider } from './chakra-provider';
import { ErrorBoundary } from './error-boundary';
import { ReactQueryProvider } from './react-query-provider';

export function Providers({ children }: { children: ReactNode }) {
  return (
    <ReactQueryProvider>
      <ChakraProvider>
        <ErrorBoundary>
          <AuthProvider>
            <BrowserRouter>{children}</BrowserRouter>
          </AuthProvider>
        </ErrorBoundary>
      </ChakraProvider>
    </ReactQueryProvider>
  );
}
