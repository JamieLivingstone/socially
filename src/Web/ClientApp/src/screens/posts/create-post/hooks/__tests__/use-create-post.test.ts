import { renderHook } from '@testing-library/react-hooks';
import moxios from 'moxios';

import { Providers } from '../../../../../providers';
import { CreatePostResponse, useCreatePost } from '../use-create-post';

describe('useCreatePost', () => {
  test('creates post and navigates to slug', async () => {
    const response: CreatePostResponse = {
      slug: 'test-post',
    };

    moxios.stubRequest('/api/v1/posts', {
      status: 201,
      response,
    });

    const { result } = renderHook(() => useCreatePost(), { wrapper: Providers });

    await result.current.mutateAsync({
      title: 'test-post',
      body: 'Test post body',
      tags: ['tag-one'],
    });

    expect(window.location.href).contains(response.slug);
  });
});
