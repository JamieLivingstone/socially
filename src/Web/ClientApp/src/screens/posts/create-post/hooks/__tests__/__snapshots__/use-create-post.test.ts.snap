// Vitest Snapshot v1

exports[`useCreatePost > creates post and navigates to slug 1`] = `"{\\"title\\":\\"test-post\\",\\"body\\":\\"Test post body\\",\\"tags\\":[\\"tag-one\\"]}"`;
